sips/sip-010/sip-010-fungible-token-standard.md

```diff
@@ -0,0 +1,120 @@
+
+### Balance of
+
+`(balance-of (principal) (response uint uint))`
+
+Return the balance of a particular principal (also known as "address" or "account"). Implementations should typically use the built-in Clarity method `ft-get-balance`.
+
+### Total supply
+
+`(total-supply () (response uint uint))`
+
+Return the total supply of this token. Implementations should typically use the built-in Clarity method `ft-get-supply`.
+
+## Trait implementation
+
+An implementation of the proposed trait is provided below.
+
+```clarity
+(define-trait ft-trait
+  (
+    ;; Transfer from the caller to a new principal
+    (transfer (principal uint) (response bool uint))
+
+    ;; the human readable name of the token
+    (name () (response (string-ascii 32) uint))
+
+    ;; the ticker symbol, or empty if none
+    (symbol () (response (string-ascii 32) uint))
+
+    ;; the number of decimals used, e.g. 6 would mean 1_000_000 represents 1 token
+    (decimals () (response uint uint))
+
+    ;; the balance of the passed principal
+    (balance-of (principal) (response uint uint))
+
+    ;; the current total supply (which does not need to be a constant)
+    (total-supply () (response uint uint))
+  )
+)
+```
+
+## Example implementation
+
+I have provided a simple implementation along with a Javascript client and tests. https://github.com/hstove/stacks-fungible-token
+
+## Credit
+
+Credit for the work behind this proposal belongs to [@psq](https://github.com/psq), who wrote almost all of this code originally for use in his projects (such as [Flexr](https://github.com/psq/flexr)). I am simply re-packaging his existing code in a format that can be shared in this proposal.
+
+## Implementing in wallets and other applications
+
+Developers who with to interact with a fungible token contract should first be provided, or keep track of, various different fungible token implementations. When validating a fungible token contract, they should fetch the interface and/or source code for that contract. If the contract implements the trait, then the wallet can use this standard's contract interface for making transfers and getting balances.
+
+### Use of post conditions
+
+In addition to fantastic built-in methods for fungible token contracts, the Stacks blockchain includes an innovation known as Post Conditions. By defining post conditions, users can create transactions that include pre-defined guarantees about what might happen in that contract.
+
+One such post condition could be "I will transfer exactly 100 of X token", where "X token" is referenced as a specific contract's fungible token. When wallets and applications implement the `transfer` method, they should _always_ use post conditions to specify that the user will transfer exactly the amount of tokens that they specify in the `amount` argument of the `transfer` function. Only in very specific circumstances should such a post condition not be included.
+
```

Comment:
The challenge with post conditions is that this will require custom work for each token to find out what the internal token name, if used.
A possible remedy would to have an other, optional trait, that would provide a function that returns a possible template that can be used to build post-conditions in an automatic fashion, if the trait is available.
The actual information that needs to be returned would require experimenting with a few tokens to figure out what is needed.
That function should be read-only as well.

Reply:
In the web wallet, I pull the token name from the contract interface RPC, so it's fortunately almost solved!
![image](https://user-images.githubusercontent.com/1109058/105792405-36fa9380-5f3c-11eb-9728-7c1342f96a41.png)

A problem with contract-defined post conditions is the potential for abuse. IMO, post conditions should be left as controlled by the wallet/user as much as possible.

Reply:
there may be things like multipliers to use compared to what `balance-of` return to implement rebasing tokens with an underlying token, at least I think it would be feasible, but you'll need hints from the contract, in some cases so you can avoid custom code per token

Reply:
The hodl token has two token assets, one is increased and one reduced on un-/hodling

Reply:
Yep I know you've thought about this more than I have, given Flexr. It seems out of scope, but a very valid thing to consider, especially since we should ensure that post conditions are used to their full advantage.

Reply:
I think to the greatest extent possible, the smart contract should not be responsible for enforcing post-conditions.  As I've said elsewhere, post-conditions are a proactive security measure -- they make it so the user does not need to trust smart contracts.

That said, contracts have free reign to use whatever token(s) they want internally, and use them however they want internally.  So, how would the user's wallet know which tokens movements to permit in post-conditions?

If we tweak https://github.com/blockstack/stacks-blockchain#1968 so that (1) it can call read/write functions and (2) it can return the internal `AssetMap` structure the node uses to track asset movement during a transaction's execution, then the wallet can do the following to generate the appropriate post-conditions:

1. Invoke the `transfer` method (or whatever method) via the node's "read-only" RESTful API to see what would happen if the user actually did call that function, and get back the list of assets that would have moved.
2. Construct an allowed list of post-conditions from the asset map, and deny all other token movements.
3. Show the user the list of assets that would be moved if the user did in fact send this transaction, and give the user an opportunity to adjust the post-conditions if need be.
4. Submit the transaction

If all goes well, the contract will process the method the same way it did in the "read-only" RESTful API call and the user's assets will be appropriately guarded.

What do you think?

Reply:
Yeah that sounds like a best of both worlds!

Reply:
Circling back to this, I don't think https://github.com/blockstack/stacks-blockchain#1968 is blocking ratification of this SIP?

Reply:
Going to go ahead and resolve this.  I think if we want a method for giving a hint on what post-conditions to use, it can be a separate trait.  Recall that contracts can implement multiple traits, implicitly or explicitly.